doc/developers/contributing.rst

```diff
@@ -1374,10 +1370,65 @@ this can be achieved with::
         D = self.decision_function(X)
         return self.classes_[np.argmax(D, axis=1)]
 
-In linear models, coefficients are stored in an array called ``coef_``,
-and the independent term is stored in ``intercept_``.
-``sklearn.linear_model.base`` contains a few base classes and mixins
-that implement common linear model patterns.
+In linear models, coefficients are stored in an array called ``coef_``, and the
+independent term is stored in ``intercept_``.  ``sklearn.linear_model.base``
+contains a few base classes and mixins that implement common linear model
+patterns.
 
 The :mod:`sklearn.utils.multiclass` module contains useful functions
 for working with multiclass and multilabel problems.
+
+Estimator Tags
+--------------
+.. warning::
+
+    The estimator tags are experimental and the API is subject to change.
+
+Scikit-learn introduced estimator tags in version 0.21.  These are annotations
+of estimators that allow programmatic inspection of their capabilities, such as
+sparse matrix support, supported output types and supported methods.  The
+estimator tags are a dictionary returned by the method ``_get_tags()``.  These
+tags are used by the common tests and the ``check_estimator`` function to
+decide what tests to run and what input data is appropriate.
+
+The default value of all tags is ``False``.
+
+The current set of estimator tags are:
+
+non_deterministic
+    whether the estimator is not deterministic given a fixed ``random_state``
+
+requires_positive_data - unused for now
+    whether the estimator requires positive X.
+
+no_validation
+    whether the estimator skips input-validation. This is only meant for stateless and dummy transformers!
+
+multioutput - unused for now
+    whether a regressor supports multi-target outputs or a classifier supports multi-class multi-output.
+
+multilabel
+    whether the estimator supports multilabel output
+
+stateless
+    whether the estimator needs access to data for fitting. Even though
+    an estimator is stateless, it might still need a call to ``fit`` for initialization.
```

Comment:
(should we deprecate the need for a call to fit for initialisation in stateless estimators?)

Reply:
I don't think we can, because "stateless" can still mean it depends on ``n_features``. ``RBFSampler`` for example needs to sample a random matrix of shape ``(n_features, n_components)``. I'm not sure how to do it unless we call ``fit``.
Do we error if ``Normalizer`` is called with different ``n_features`` during transform? do we want to?

Reply:
I'm not sure "stateless" is the right word then? We mean "data independent"?

Reply:
do we want to have separate tags for data independent and no state at all?

Reply:
> do we want to have separate tags for data independent and no state at all?

Dunno. What's the use case? If so, we could consider a ternary tag...

Reply:
I guess the main use-case here was that some estimators didn't complain if the number of features was different in fit and transform, possibly only AdditiveChi2Sampler.
Though this estimator still required calling fit.
I guess we can define stateless as "doesn't require calling fit" and everything that requires calling fit should check the shape?

Reply:
Hm AdditiveChi2Sampler requires calling ``fit`` for no reason actually...

Reply:
Opened #12616 to follow up. I don't think there's a good reason for a ternary tag. Right now this is used for testing two things: checking that calling ``transform`` before ``fit`` will raise a nice error, and checking that the number of features needs to be consistent between ``fit`` and ``transform``.
These checks are somewhat independent, but both related to being stateless.